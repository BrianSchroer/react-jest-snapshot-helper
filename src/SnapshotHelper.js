import React from 'react';
import renderer from 'react-test-renderer';

/**
 * React component Jest snapshot helper
 */
export default class SnapshotHelper {
  /**
   * Create new SnapshotHelper instance
   * @param {*} component - React component
   */
  constructor(component) {
    this.component = component.type;
    this.props = component.props;
    this.adjustProps = (props, propsAdjustments) =>
      Object.assign({}, props, propsAdjustments);
  }

  /**
   * Fluent syntax helper to specify a callback function to be called by the test() function
   * to assign propsAdjustments to the props passed that were passed via the constructor.
   * (Defaults to assigning top-level props only.)
   */
  withPropsAdjuster = propsAdjuster => {
    this.adjustProps = propsAdjuster;
    return this;
  };

  /**
   * Adjust props that were passed to the constructor with propsAdjustments,
   * then create JSON string snapshot for the component.
   */
  getSnapshot = propsAdjustments => {
    const Component = this.component;
    const props = this.adjustProps(this.props, propsAdjustments);

    return renderer.create(<Component {...props} />).toJSON();
  };

  /**
   * Adjust props that were passed to the constructor with propsAdjustments, then create
   * snapshot and assert that it matches previously accepted screenshot.
   */
  test = propsAdjustments => {
    expect(this.getSnapshot(propsAdjustments)).toMatchSnapshot();
  };

  /**
   * Get JSON string snapshot for the component.
   * @param {*} component - React component
   */
  static getSnapshot(component) {
    return new SnapshotHelper(component).getSnapshot();
  }

  /**
   * Create snapshot and assert that it matches previously accepted screenshot.
   * @param {*} component - React component1
   */
  static test(component) {
    new SnapshotHelper(component).test();
  }
}
