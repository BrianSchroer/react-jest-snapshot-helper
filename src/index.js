export { default as SnapshotHelper } from './SnapshotHelper';
