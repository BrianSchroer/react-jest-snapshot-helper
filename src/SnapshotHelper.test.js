import React from 'react';
import renderer from 'react-test-renderer';
import { SnapshotHelper } from '../src';

/* eslint-disable react/prop-types */

const TestComponent = ({ className, text, details }) => (
  <div className={className}>
    <p>{text}</p>
    <p>{details.description}</p>
  </div>
);

describe('SnapshotHelper', () => {
  const testComponentInstance = (
    <TestComponent
      className="test-class-name"
      text="test text"
      details={{ description: 'detail description' }}
    />
  );

  const defaultProps = testComponentInstance.props;
  const snapshotHelper = new SnapshotHelper(testComponentInstance);

  it('static .getSnapshot() should return expected JSON', () => {
    const expected = renderer.create(testComponentInstance).toJSON();
    const actual = SnapshotHelper.getSnapshot(testComponentInstance);
    expect(actual).toEqual(expected);
  });

  it('static .test() should save expected snapshot', () => {
    SnapshotHelper.test(testComponentInstance);
  });

  it('.getSnapshot() with no overrides should return expected JSON', () => {
    const expected = renderer.create(testComponentInstance).toJSON();
    const actual = snapshotHelper.getSnapshot();
    expect(actual).toEqual(expected);
  });

  it('.getSnapshot() with overrides should return expected JSON', () => {
    const instance = (
      <TestComponent
        className="different-class-name"
        text="different text"
        details={{ description: 'different description' }}
      />
    );
    const expected = renderer.create(instance).toJSON();

    const actual = snapshotHelper.getSnapshot({
      className: 'different-class-name',
      text: 'different text',
      details: { description: 'different description' }
    });

    expect(actual).toEqual(expected);
  });

  it('.getSnapshot() with overrides should not permanently update component from constructor', () => {
    const snapshotBefore = snapshotHelper.getSnapshot();

    snapshotHelper.getSnapshot({ text: 'different text' });

    const snapshotAfter = snapshotHelper.getSnapshot();
    expect(snapshotAfter).toEqual(snapshotBefore);
  });

  it('.test() with no overrides should save expected snapshot', () => {
    snapshotHelper.test();
  });

  it('.test() with overrides should not permanently update component from constructor', () => {
    const snapshotBefore = snapshotHelper.getSnapshot();

    snapshotHelper.test({
      className: 'different-class-name',
      text: 'different text',
      details: { description: 'different description' }
    });

    const snapshotAfter = snapshotHelper.getSnapshot();
    expect(snapshotAfter).toEqual(snapshotBefore);
  });

  describe('withPropsAdjuster', () => {
    const adjustedHelper = new SnapshotHelper(
      <TestComponent {...defaultProps} />
    ).withPropsAdjuster((props, adjustments) => ({
      className: defaultProps.className,
      text: defaultProps.text,
      details: Object.assign({}, props.details, adjustments)
    }));

    it('.withPropsAdjuster().getSnapshot() with no overrides should return expected JSON', () => {
      const expected = renderer.create(testComponentInstance).toJSON();
      const actual = adjustedHelper.getSnapshot();
      expect(actual).toEqual(expected);
    });

    it('.withPropsAdjuster().getSnapshot() with overrides should return expected JSON', () => {
      const instance = (
        <TestComponent
          className={defaultProps.className}
          text={defaultProps.text}
          details={{ description: 'different description' }}
        />
      );

      const expected = renderer.create(instance).toJSON();

      const actual = adjustedHelper.getSnapshot({
        description: 'different description'
      });

      expect(actual).toEqual(expected);
    });
  });
});
