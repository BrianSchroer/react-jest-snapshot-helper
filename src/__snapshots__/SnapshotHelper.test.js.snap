// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnapshotHelper .test() with no overrides should save expected snapshot 1`] = `
<div
  className="test-class-name"
>
  <p>
    test text
  </p>
  <p>
    detail description
  </p>
</div>
`;

exports[`SnapshotHelper .test() with overrides should not permanently update component from constructor 1`] = `
<div
  className="different-class-name"
>
  <p>
    different text
  </p>
  <p>
    different description
  </p>
</div>
`;

exports[`SnapshotHelper static .test() should save expected snapshot 1`] = `
<div
  className="test-class-name"
>
  <p>
    test text
  </p>
  <p>
    detail description
  </p>
</div>
`;
